Setting up MongoDB Integration

farhan0167 opened the pull request:
## What
This PR seeks to integrate MongoDB with Vectorflow
## Why
Currently Vectorflow does not have a mongo connector, and should address #34 
## How
MongoDB as a database let's users store vector embeddings but is different from other vector databases, where in Mongo, one needs to associate a Search Index to a collection. This means that we'll be uploading our embeddings to a Mongo databases' collection, and that users of Mongo are expected to create their own search indexes which is associated with the collection. Since we cannot directly store our data in the Search Index, we store data in a collection as:

```python
...
def write_embeddings_to_mongodb(upsert_list, vector_db_metadata):
    mongo_conn_uri = os.getenv('VECTOR_DB_KEY')
    try:
        mongo_client = pymongo.MongoClient(mongo_conn_uri)
    except Exception as e:
        logging.error(e)
        return None
    db_name, collection = vector_db_metadata.index_name.split(".")
    db = mongo_client[db_name]

    if collection not in db.list_collection_names():
        logging.error(f"Index {vector_db_metadata.index_name} does not exist in environment {vector_db_metadata.environment}")
        return None
    
    index = db.get_collection(collection)
...
```
A few things to note:
1. To connect to Mongo, we use PyMongo's Driver `mongo_client = pymongo.MongoClient("<mongo_conn_uri>")`. This means for this implementation, the `VECTOR_DB_KEY` is the connection uri. 
2. Since we cannot specify the search index directly and we have to store embeddings in a collection, we pass a string `mongo_databasename.collection_name` to the ` vector_db_metadata.index_name`. The tests below were conducted with `databasename=vector_embeds` and `collection_name=test`.
3. While adding `pymongo` certain dependencies in the `requirements.txt` changed.
4. In the code block for connecting to Mongo, when user enters the a wrong `VECTOR_DB_KEY`, vectorflow seems to get stuck in `BATCH_PROCESSING` state even though the expected behavior is to exit with a None. Not sure if this is desirable

## Verification

Prior to data upload:
![Screenshot 2023-10-18 at 11 05 38 PM](https://github.com/dgarnitz/vectorflow/assets/44986467/50432e55-0dfe-47ac-b9bb-ada6a7745d5a)
Request to vectorflow:
![Screenshot 2023-10-18 at 11 14 20 PM](https://github.com/dgarnitz/vectorflow/assets/44986467/87643a29-94f0-4b06-b5e0-da86109bf419)
After data upload on MongoDB:
![Screenshot 2023-10-18 at 11 14 47 PM](https://github.com/dgarnitz/vectorflow/assets/44986467/824e007c-6145-4786-ac91-8e28ac5616bf)

### Unit Test Results:
![Screenshot 2023-10-18 at 11 19 48 PM](https://github.com/dgarnitz/vectorflow/assets/44986467/d9397424-551a-4526-8e26-66d7927cf4f8)
and
![Screenshot 2023-10-18 at 11 20 36 PM](https://github.com/dgarnitz/vectorflow/assets/44986467/8c40d1ff-484a-4f2f-8ddc-e4ade3846351)
The last test failure, as agreed, is due an existing issue not related to Mongo integration, so I'm considering that all test cases have passed.

farhan0167 commented:
@dgarnitz, let me know what you think of the changes, and the notes I made

dgarnitz commented on src/worker/requirements.txt:
```diff
@@ -10,25 +10,27 @@ cachetools==5.3.1
 grpc-gateway-protoc-gen-openapiv2==0.1.0
-grpcio
-grpcio-tools
+grpcio==1.56.0
```
There is a weird quirk where `grpcio` and `grpcio-tools` need to have their versions undefined to avoid a dependency conflict. Please remove the version numbers from these two items.

dgarnitz commented on src/worker/requirements.txt:
```diff
@@ -10,25 +10,27 @@ cachetools==5.3.1
 hyperframe==6.0.1
 idna==3.4
-lancedb
+lancedb==0.3.1
```
same thing here with `lancedb` - please remove the version numbers to avoid a dependency clash

dgarnitz commented on src/worker/requirements.txt:
```diff
@@ -49,8 +51,9 @@ pyarrow==13.0.0
 pydantic==2.1.1
 pydantic_core==2.4.0
-pylance
+pylance==0.8.3
```
same thing here with`pylance` - please remove the version numbers to avoid a dependency clash

dgarnitz reviewed (changes requested):
This is fantastic work! I would appreciate if you could please rebase the latest changes from master since I just merged a big PR and retest the ingestion one last time. Also can you please make the changes to the requirements.txt? thanks!!

dgarnitz commented on src/worker/vdb_upload_worker.py:
```diff
@@ -70,9 +71,57 @@ def write_embeddings_to_vector_db(text_embeddings_list, vector_db_metadata, batc
+    try:
+        mongo_client = pymongo.MongoClient(mongo_conn_uri)
+    except Exception as e:
+        logging.error(e)
```
Can you alter this logging statement to have a bit more information in it? Something like:
`logging.error('Error connecting to MongoDB via python client: %s', e)`

dgarnitz commented:
@farhan0167 with regards to your comment:
"_In the code block for connecting to Mongo, when user enters the a wrong VECTOR_DB_KEY, vectorflow seems to get stuck in BATCH_PROCESSING state even though the expected behavior is to exit with a None. Not sure if this is desirable_"

This is not desirable. The system needs to mark the batch as failed. Did you determine if it is actually returning `None` and just failing to update, or is it not reaching the `return None` part of the code? 

If its not reaching that part of the code, is it because the system hangs trying to connect to the DB over and over again? If so, I'm pretty sure the MongoDB client will have some configuration that can tell it to only retry X times before failing or some type of timeout that you can set to like 10 seconds.

farhan0167 commented:
@dgarnitz I'll get to it. Also can you clarify one thing as well. What are your thoughts on the implementation where we need to pass the mongodb connection uri to the `mongo_conn_uri = os.getenv('VECTOR_DB_KEY')` as a `VECTOR_DB_KEY`? I was looking at the documentation last night, and I noticed that you've added uri's in the `vector_db_metadata.environment` previously. What purpose would `environment` serve for this implementation since I believe there is no unique environment variable that needs to be passed for Mongo. When I was testing it, I arbitrarily passed an environment variable `mongo` to it, but it does not serve any purpose in the code.
![Screenshot 2023-10-20 at 2 12 15 PM](https://github.com/dgarnitz/vectorflow/assets/44986467/31464e54-fd7a-4872-9549-1275ac030667)

dgarnitz commented:
@farhan0167 if the connection URI has a piece that is the database password, pass that as the `VECTOR_DB_KEY` and the rest of it as the environment variable. The reason is that the environment value is stored in the DB whereas the key is not, so its more secure

farhan0167 commented:
@dgarnitz, I've resolved the issues you brought up, and below are some notes on the changes:

1. Rebased with the most recent changes pushed.
2. Removed the dependency versions from the requested packages
3. Fixed the part where previously inserting a wrong connection string kept the system in `BATCH PROCESSING` status instead of `FAILED`. The issue was in the way MongoDB connections work as referenced [here](https://pymongo.readthedocs.io/en/stable/api/pymongo/mongo_client.html#:~:text=URIs%20are%20used.-,Note,-Starting%20with%20version). This essentially means that, for us to validate whether the connection string we passed succeeded was by pinging a database within that account.
4. As for the `VECTOR_DB_KEY` issue, users will need to pass the connection string to the environment variable as 
``` 
mongodb+srv://<username>:<password>@cluster0.zin0tgq.mongodb.net/?retryWrites=true&w=majority
```
 The user needs to modify the `<username>` with their username in the environment field, and pass the `<password>` as `VECTOR_DB_KEY`. The connection string is usually retrieved from the `Atlas Console->Databases->Connect->Drivers->Python` as:
![Screenshot 2023-10-24 at 1 14 01 AM](https://github.com/dgarnitz/vectorflow/assets/44986467/7d138388-5c1a-48dd-8c1c-7a5299788f6a).
5. Reran all the tests with success:
![Screenshot 2023-10-24 at 12 29 02 AM](https://github.com/dgarnitz/vectorflow/assets/44986467/bd64640c-00f8-4c76-9428-53ffe8d49eba)
Additionally, in terms of success with uploading to Mongo, everything works and uploads as expected like before.

dgarnitz reviewed (approved):
Excellent work!!
